### ISSUE TITLE
Broke compatibility with PlatformIO

### ISSUE THREAD

**Issue by dnk1000** [Microsoft/vscode#47350]
Issue Type: <b>Bug</b>

Try and compile a program for ESP-IDF using Platform IO

VS Code version: Code 1.22.1 (950b8b0d37a9b7061b6f0d291837ccc4015f5ecd, 2018-04-06T02:26:57.615Z)
OS version: Windows_NT x64 6.1.7601

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz (4 x 2808)|
|Memory (System)|7.86GB (3.36GB free)|
|Process Argv|C:\Program Files\Microsoft VS Code\Code.exe|
|Screen Reader|no|
|VM|0%|

</details><details><summary>Extensions (7)</summary>

Extension|Author (truncated)|Version
---|---|---
code-gnu-global|aus|0.2.2
vscode-github|Kni|0.27.1
python|ms-|2018.3.1
cpptools|ms-|0.16.1
platformio-ide|pla|0.14.0
vscode-icons|rob|7.22.0
debug|web|0.22.0

(3 theme extensions excluded)

</details>

**Comment by gschintgen** [comment]
Same here. The auto-update of Visual Studio Code broke my platformio setup. There's also a report on platformio's tracker with some more details:
https://github.com/platformio/platformio-vscode-ide/issues/97
This is a serious regression, I just don't know if it's a bug in vscode or platformio.

**Comment by roblourens** [comment]
- Do you have a .vscode/tasks.json file? Can you post it?
- In the upstream issue it shows the terminal writing a path to platformio.exe. Is the path correct?

**Comment by gschintgen** [comment]
There's no tasks.json file in my project's .vscode folder. (Is that to be expected?)
No, the path to platformio.exe is not correct and AFAICT it should not have been added to the command. I.e. simply calling "platformio run" (without specifying the path to the .exe) in the project's source folder works just fine if I open the integrated PlatformIO Terminal.
I can't remember if the previous version of vscode (probably 1.21) also added a (correct) path to the .exe or if it simply relied on the .exe being found through the PATH variable.

**Comment by roblourens** [comment]
That's ok, it sounds like the extension is contributing a task and either vscode is interpreting it incorrectly or the task is not formed correctly. @dbaeumer or the extension maintainer can investigate more.

**Comment by ivankravets** [comment]
@Tyriar is this bug related to Terminal?

**Comment by ivankravets** [comment]
A temporary solution is to roll back to 1.21 using https://code.visualstudio.com/updates/v1_21 and "Download" links.

Also, maybe you will need to disable automatic updates for VSCode.

**Comment by ivankravets** [comment]
Yes, the only Windows is affected.

**Comment by dbaeumer** [comment]
I am able to reproduce this. Will look into what broke it.

**Comment by dbaeumer** [comment]
Under Windows I wrote my own code to find the executable on the path. This got broken with https://github.com/Microsoft/vscode/commit/b1adb2179631ff83ab6bb1e43bc343e2835cd985 which fixed https://github.com/Microsoft/vscode/issues/39720

**Comment by dbaeumer** [comment]
I started to debug why this is broken and the problem is that I when VS Code tries to find the executable the C:\Users\dirkb\.platformio\penv\Scripts is not on the PATH in my system (e.g. the renderer) but it is in the terminal. Need to understand why this is the case.

**Comment by dbaeumer** [comment]
Understand it now. The reason for the breakage is that Platform IO brings its own PATH environment and `findExecutable` doesn't consider this. It only takes the PATH from the current renderer process.

**Comment by dbaeumer** [comment]
The interesting questions is why that worked in the first place since under Windows the OS doesn't try to find an executable on the PATH when launching a process. The PATH must always be absolute. 

@Tyriar did the terminal or some of the underlying libraries get code to find the executable on the PATH under Windows. This would explain why it worked before. If this is the case then I could simply remove my code :-)

**Comment by dbaeumer** [comment]
OK. I found the code: https://github.com/Tyriar/node-pty/blob/master/src/win/path_util.cc#L31

However that only works if the extension is already provided. So I need to keep my code since simply using `platformio` would have been enough. The task system then looks for `platformio.com` and `platformio.exe`

**Comment by dbaeumer** [comment]
To verify:

- install Platform IO (see http://docs.platformio.org/en/latest/ide/vscode.html)
- create a Project (http://docs.platformio.org/en/latest/ide/vscode.html#quick-start)
- Run Task > PlatformIO: Build

The build task should succeed.

**Comment by dbaeumer** [comment]
Reopening the issue to discussion in bug standup.

**Comment by Tyriar** [comment]
> OK. I found the code

Yes that's it, there's an open issue to improve this: https://github.com/Tyriar/node-pty/issues/112

**Comment by dbaeumer** [comment]
@Tyriar thanks. I wrote which in TS if that would help for you. It is here: https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/tasks/electron-browser/terminalTaskSystem.ts#L691

**Comment by dbaeumer** [comment]
We decided to include the fix in the fix in the next recovery build.

**Comment by Tyriar** [comment]
@dbaeumer that does indeed help :+1: is this something we can push to node-pty and rely on it there?
